arch/arm/core/aarch32/prep_c.c

```diff
@@ -40,43 +40,53 @@
 #ifdef CONFIG_CPU_CORTEX_M_HAS_VTOR
 
 #ifdef CONFIG_XIP
-#define VECTOR_ADDRESS ((uintptr_t)_vector_start)
+#ifdef CONFIG_SW_VECTOR_RELAY
+#define VECTOR_ADDRESS ((uintptr_t)__vector_relay_table)
 #else
-#define VECTOR_ADDRESS CONFIG_SRAM_BASE_ADDRESS
+#define VECTOR_ADDRESS ((uintptr_t)_vector_start)
 #endif
-static inline void relocate_vector_table(void)
-{
-	SCB->VTOR = VECTOR_ADDRESS & SCB_VTOR_TBLOFF_Msk;
-	__DSB();
-	__ISB();
-}
+#else /* CONFIG_XIP */
+#define VECTOR_ADDRESS CONFIG_SRAM_BASE_ADDRESS
+#endif /* CONFIG_XIP */
 
-#else
+#else /* CONFIG_CPU_CORTEX_M_HAS_VTOR */
+
+#define VECTOR_ADDRESS 0
+
+#endif /* CONFIG_CPU_CORTEX_M_HAS_VTOR */
 
 #if defined(CONFIG_SW_VECTOR_RELAY) || defined(CONFIG_SW_VECTOR_RELAY_CLIENT)
 Z_GENERIC_SECTION(.vt_pointer_section) __attribute__((used))
 void *_vector_table_pointer;
 #endif
 
-#define VECTOR_ADDRESS 0
 
+#ifdef CONFIG_CPU_CORTEX_M_HAS_VTOR
+static inline void relocate_vector_table(void)
+#else
 void __weak relocate_vector_table(void)
+#endif
 {
+#if defined(CONFIG_CPU_CORTEX_M_HAS_VTOR) && !defined(CONFIG_SW_VECTOR_RELAY_CLIENT)
```

Comment:
It's a bit difficult to see the use-case here: if I have VTOR, why do I need to be a chain-loadable leaf firmware client? I can just make VTOR point to my vector table.

Reply:
To put it differently: in line 46 you set VECTOR_ADDRESS for the case where HAS_VTOR and RELAY_CLIENT are both set, but this value is not used anywhere.

Reply:
I agree with wondering what the reason is to be able to have both VTOR and the relay feature enabled?